plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,484 @@
+                    }
+            sendContractInboxMessage @t (UtxoAtResponse response) i it
+    in fold <$> traverse go (Set.toList addresses)
+
+processWriteTxRequests ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member ChainIndexEffect effs
+    , Member WalletEffect effs
+    , Member Log effs
+    , Member SigningProcessEffect effs
+    )
+    => Eff effs ()
+processWriteTxRequests = do
+    logInfo "processWriteTxRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.writeTxRequests @t)
+    unless (Map.null state) $ do
+        traverse_ (uncurry $ processWriteTxRequest @t) (Map.toList state)
+    logInfo "processWriteTxRequests start"
+
+processWriteTxRequest ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member Log effs
+    , Member ChainIndexEffect effs
+    , Member WalletEffect effs
+    , Member SigningProcessEffect effs
+    )
+    => ContractInstanceId
+    -> (ContractIteration, PendingTransactions)
+    -> Eff effs [ChainEvent t]
+processWriteTxRequest i (it, PendingTransactions txns) =
+    let go unbalancedTx = do
+            logInfo "processWriteTxRequest"
+            logInfo "Start watching contract addresses."
+            AM.addressesTouched <$> watchedAddresses <*> pure (unBalancedTxTx unbalancedTx) >>= traverse_ startWatching
+            logInfo $ "Balancing unbalanced TX: " <> tshow unbalancedTx
+            r <- runError $ do
+                    balancedTx <- balanceWallet unbalancedTx
+                    signedTx <- signWithOwnPublicKey balancedTx
+                    logInfo $ "Storing signed TX: " <> tshow signedTx
+                    void $ runCommand (saveBalancedTx @t) WalletEventSource balancedTx
+                    logInfo $ "Submitting signed TX: " <> tshow signedTx
+                    balanceResult <- submitTx signedTx
+                    void $ runCommand (saveBalancedTxResult @t) NodeEventSource balanceResult
+                    pure balanceResult
+            let response = either WriteTxFailed (WriteTxSuccess . Ledger.txId) r
+            logInfo . render $ "processWriteTxRequest result:" <+> pretty response
+            sendContractInboxMessage (WriteTxResponse response) i it
+    in fold <$> traverse go txns
+
+-- | A wrapper around the NodeAPI function that returns some more
+-- useful evidence of the work done.
+submitTx :: (Member WalletEffect effs) => Ledger.Tx -> Eff effs Ledger.Tx
+submitTx tx = submitTxn tx >> pure tx
+
+processNextTxAtRequests ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member Log effs
+    )
+    => Eff effs ()
+processNextTxAtRequests = do
+    logInfo "processNextTxAtRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.nextTxAtRequests @t)
+    traverse_ (uncurry $ processNextTxAtRequest @t) (Map.toList state)
+    logInfo "processNextTxAtRequests end"
```

Comment:
I wonder if we should extract this pattern into a function so you can do
```
logOperation "whatever" $ do ...
```
and it does the start/end messages.

Reply:
Yeah I want to do that as part of the "improve logging/debugging" task